src/compiler.ts

```diff
@@ -4620,8 +4622,50 @@ export class Compiler extends DiagnosticEmitter {
         let targetType = leftType;
         let instance: Function | null;
 
-        // Mathf.pow if lhs is f32 (result is f32)
-        if (this.currentType.kind == TypeKind.F32) {
+        if ((this.currentType.kind == TypeKind.I32 || this.currentType.kind == TypeKind.U32)) {
```

Comment:
I think this should handle all kinds of integers, `this.currentType.is(TypeFlags.INTEGER | TypeFlags.VALUE)`, including the small ints?

Reply:
Yeah. But I got this validation errors:
```
Testing compiler/resolve-binary

[wasm-validator error in function start:resolve-binary] call param types must match, on 
[i32] (call $~lib/number/F64#toString
 [i32] (call $~lib/math/ipow32
  [i32] (i32.const 2)
  [i32] (i32.const 2)
 )
 [i32] (i32.const 0)
)
(on argument 0)

  Error: Validate error
- Created fixture

[wasm-validator error in function start:resolve-binary] call param types must match, on 
[i32] (call $~lib/number/F64#toString
 [i32] (call $~lib/math/ipow32
  [i32] (i32.const 2)
  [i32] (i32.const 2)
 )
 [i32] (i32.const 0)
)
(on argument 0)

```

Reply:
This looks like the result type of the generated instructions is set to `Type.f32` instead of one of the integer types. Must set `this.currentType = theOutputTypeOfTheEmittedInstruction` (here of `call $ipow32`).

Reply:
I found comment `// TODO: should the result be converted back?`. I guess if `type size < 32` bit we need convert it back after `ipow32()` call. How better to do this?

Reply:
Pretty interesting. This issue solved via expression:
```ts
 if (
   this.currentType.kind == TypeKind.I8  ||
   this.currentType.kind == TypeKind.U8  ||
   this.currentType.kind == TypeKind.I16 ||
   this.currentType.kind == TypeKind.U16 ||
   this.currentType.kind == TypeKind.I32 ||
   this.currentType.kind == TypeKind.U32
) {
  ...
}
```

Reply:
Would say the comment becomes irrelevant with `ipow` in place. Was for `i32 -> f32 -> pow -> X`, but in the `ipow32` case there shouldn't be any conversion: `i32 -> ipow32 -> still i32`. Perhaps that function needs a new switch structure now, doing

i8, i16, i32, isize (Wasm32) -> ipow32
u8, u16, u32, usize (Wasm32) -> upow32?
i64 -> ipow64
u64 -> upow64?
f32 -> powf
f64 -> pow

according to the input type?

Reply:
@dcodeIO Plz see last commit. Commented parts cause to that issue.

Reply:
It seems for small types we need post operation wrapping